File: menu-ocr-ios/MenuOCR/MenuOCR/ViewModels/PaymentViewModel.swift
//
//  PaymentViewModel.swift
//  MenuOCR
//
//  Created by MenuOCR on 2025-01-07.
//

import Foundation
import Combine

class PaymentViewModel: ObservableObject {
    @Published var paymentState: PaymentState = .idle
    @Published var paymentHistoryState: PaymentHistoryState = .idle

    private let apiService = ApiService()
    private var cancellables = Set<AnyCancellable>()

    func createPaymentIntent(amount: Int, description: String? = nil) {
        paymentState = .loading

        let request = PaymentIntentRequest(
            amount: amount,
            description: description
        )

        Task {
            do {
                let response = try await apiService.createPaymentIntent(request: request)
                await MainActor.run {
                    self.paymentState = .paymentIntentCreated(response)
                }
            } catch {
                await MainActor.run {
                    self.paymentState = .error(error.localizedDescription)
                }
            }
        }
    }

    func loadPaymentHistory() {
        paymentHistoryState = .loading

        Task {
            do {
                let response = try await apiService.getPaymentHistory()
                await MainActor.run {
                    self.paymentHistoryState = .success(response)
                }
            } catch {
                await MainActor.run {
                    self.paymentHistoryState = .error(error.localizedDescription)
                }
            }
        }
    }

    func presentPaymentSheet(paymentIntent: PaymentIntentResponse, from viewController: UIViewController) {
        PaymentService.shared.presentPaymentSheet(
            paymentIntentClientSecret: paymentIntent.client_secret,
            from: viewController
        ) { [weak self] result in
            guard let self = self else { return }

            Task { @MainActor in
                switch result {
                case .completed:
                    self.paymentState = .paymentCompleted
                case .canceled:
                    self.paymentState = .paymentCanceled
                case .failed(let error):
                    self.paymentState = .error(error.localizedDescription)
                }
            }
        }
    }
}

enum PaymentState {
    case idle
    case loading
    case paymentIntentCreated(PaymentIntentResponse)
    case paymentCompleted
    case paymentCanceled
    case error(String)
}

enum PaymentHistoryState {
    case idle
    case loading
    case success(PaymentHistoryResponse)
    case error(String)
}

File: menu-ocr-android/app/src/main/java/com/menuocr/viewmodel/PaymentViewModel.kt
package com.menuocr.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.menuocr.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class PaymentViewModel @Inject constructor(
    private val apiService: ApiService,
    private val paymentService: PaymentService
) : ViewModel() {

    private val _paymentState = MutableStateFlow<PaymentState>(PaymentState.Idle)
    val paymentState: StateFlow<PaymentState> = _paymentState

    private val _paymentHistoryState = MutableStateFlow<PaymentHistoryState>(PaymentHistoryState.Idle)
    val paymentHistoryState: StateFlow<PaymentHistoryState> = _paymentHistoryState

    fun createPaymentIntent(amount: Int, description: String? = null) {
        viewModelScope.launch {
            _paymentState.value = PaymentState.Loading
            try {
                val request = PaymentIntentRequest(
                    amount = amount,
                    description = description
                )
                val response = apiService.createPaymentIntent(request)

                if (response.isSuccessful) {
                    response.body()?.let { paymentIntent ->
                        _paymentState.value = PaymentState.PaymentIntentCreated(paymentIntent)
                    } ?: run {
                        _paymentState.value = PaymentState.Error("Failed to create payment intent")
                    }
                } else {
                    _paymentState.value = PaymentState.Error("Failed to create payment intent: ${response.message()}")
                }
            } catch (e: Exception) {
                _paymentState.value = PaymentState.Error(e.message ?: "Unknown error")
            }
        }
    }

    fun loadPaymentHistory() {
        viewModelScope.launch {
            _paymentHistoryState.value = PaymentHistoryState.Loading
            try {
                val response = apiService.getPaymentHistory()

                if (response.isSuccessful) {
                    response.body()?.let { history ->
                        _paymentHistoryState.value = PaymentHistoryState.Success(history)
                    } ?: run {
                        _paymentHistoryState.value = PaymentHistoryState.Error("Failed to load payment history")
                    }
                } else {
                    _paymentHistoryState.value = PaymentHistoryState.Error("Failed to load payment history: ${response.message()}")
                }
            } catch (e: Exception) {
                _paymentHistoryState.value = PaymentHistoryState.Error(e.message ?: "Unknown error")
            }
        }
    }

    fun presentPaymentSheet(paymentIntent: PaymentIntentResponse) {
        // This would be called from the Activity/Fragment to present the PaymentSheet
        // Implementation depends on how you structure the UI
    }
}

sealed class PaymentState {
    object Idle : PaymentState()
    object Loading : PaymentState()
    data class PaymentIntentCreated(val paymentIntent: PaymentIntentResponse) : PaymentState()
    data class Error(val message: String) : PaymentState()
}

sealed class PaymentHistoryState {
    object Idle : PaymentState()
    object Loading : PaymentState()
    data class Success(val history: PaymentHistoryResponse) : PaymentState()
    data class Error(val message: String) : PaymentState()
}